### ISSUE TITLE
sassc-ruby dependency breaks JRuby+Windows support

### ISSUE THREAD

**Issue by headius** [rails/rails#38492]
### Steps to reproduce

* Generate an app on Windows using JRuby and recent Rails 6 (I think 6.0.1 introduced the sassc depenedency)
* The app will fail to bundle because it tries to compile libsass

### Expected behavior

The app should generate and bundle sucessfully.

### Actual behavior

The sassc gem contains libsass, which is built at install time and bound using FFI. **This works correctly in JRuby** when running on a platform that can build the library.

On Windows, most users do not have a C compiler, so the sassc gem includes pre-built binary gems.

Unfortunately, this **does not work on JRuby**. Because JRuby's RUBY_PLATFORM is always "java" (out of necessity, since that's the only way we can qualify JRuby extension gems), none of the binary gems are installed. RubyGems proceed to install the bare gem which tries to compile libsass, and without build tools it fails.

### System configuration
**Rails version**: I believe Rails 6.0.1 changed the dependency from "sass" to "sassc". The commit is here: 1953e0e1ace0a1790e8fe5ee0bca159637d67f61

**Ruby version**: Any JRuby 9.2.x.

### Notes

This is not strictly a Rails issue, but Rails did change the dependency in a minor release from a library that works fine on JRuby+Windows to a library that does not.

The real issue is that the sassc gem does not ship a version of the gem that will install properly on JRuby+Windows without a C++ compiler present. I will file an issue with them.

**Comment by headius** [comment]
My bad, ignore the rogue reference to a bundler issue.

The short term fix for this would possibly be to use sass on JRuby (or on JRuby on Windows, if that's possible). It's arguable that such a major dependency change should not have been done in a minor release, but you do you.

Long term the issue that needs to be fixed is sass/sassc-ruby#182, which can be fixed by adding a `-java` gem that selects the right prebuilt Windows binary. No other changes to sassc or rails would be necessary.

**Comment by rafaelfranca** [comment]
Thank you for issue but this is not an rails issue. Sass is not supported anymore and this is why rails 6 will only use sassc https://www.sass-lang.com/blog/ruby-sass-is-unsupported. If JRuby applications need to use the non-supported sass they can define that in the application gemfile.

**Comment by rafaelfranca** [comment]
Oops. Closed by mistake. We can relax the gemfile, but new apps will use sass-rails 6.0.0 anyway because bundler always resolve to the latest release

**Comment by headius** [comment]
Actually this is not just a problem for JRuby. It sounds like sassc does not work on Windows *even on CRuby* because it's trying to load the native library with the wrong extension. It is also broken on MacOS for TruffleRuby since they build libraries with a different extension that ".bundle".

See sass/sassc-ruby#173 and https://github.com/sass/sassc-ruby/issues/182#issuecomment-587662916.

Perhaps you can clarify for me what the policy is on supporting Rails on Windows? Apparently it has not worked for anyone since 6.0.1?

Agreed that this isn't strictly a Rails bug, but I'm not sure what to call it when a minor release breaks an entire platform for both CRuby and JRuby.

**Comment by headius** [comment]
Also it seems some facts were lost along the way. JRuby **does not require the sass gem** because the sassc gem works just fine if it's able to build the native library. The broken scenarios are:

* JRuby on Windows, because the binary gems will not install 
* All Rubies on Windows, because the binary gems still won't load the right file extension
* Any Ruby on MacOS if they do not build the library with a ".bundle" extension

I mentioned this above but clarify it again now: the simple fix would be to make Windows (and possibly MacOS) generate apps using the sass gem until sassc loads properly. The longer-term fix requires work on sassc to install the binaries on JRuby and choose the right extension on all Ruby impls.

**Comment by linojon** [comment]
This issue still exists today,  jruby (9.2.13.0) and Rails (6.0.3.4), on Windows 10

**Comment by headius** [comment]
I am not certain, but I think this also affects CRuby on Windows since it can't install the proper extension.

**Comment by TheSynthMaster** [comment]
Same issue here, it's supposed to be fixed or it's a lost cause? Should we abandon windows?

**Comment by mohits** [comment]
hi @headius - is this the problem?
https://notepad.onghu.com/2021/ruby-tricks-009-jruby-sassc-installation-error/

**Comment by chadlwilson** [comment]
> hi @headius - is this the problem? https://notepad.onghu.com/2021/ruby-tricks-009-jruby-sassc-installation-error/

FWIW I tried the approach in this blog which was initially very helpful, thanks! It was fine for Windows 10/11 but a pain to try to get to work for Windows Server (e.g for some kind of automated build/test in a container) and I eventually gave up.

The only practical approach I found (short of getting rid of sprockets completely) that works across JRuby+rails+sprockets was to move to https://github.com/ntkme/sassc-embedded-shim-ruby - can confirm that this works within JRuby on Mac, Linux, Windows, and also has the benefit of moving you towards a currently supported sass/scss implementation, rather than the weird limbo quasi-deprecation of sassc/libsass currently. It's not "clean" since it patches sassc-ruby to make the change transparent, but personally I felt it a reasonable trade-off to make for productivity. The shim downloads native binaries appropriate for the "real" platform (rather than Ruby-platform), and I worked with the author to fix some niggles specific to JRuby and Windows bundle installs - seems to be working well now.
